### PR TITLE
integrate supervisor into code

### LINKED ISSUE #1959: avoid confirmation dialog on restarting a workspace
Steps to reproduce:
- start a workspace
- make dirty changes and don't save them
- rearrange layout
- stop the workspace and click the start workspace button when the workspace is stopped
- there should not be any dialogs since the workspace is gone and dirty changes cannot be saved already
- on reload though layout changes should be restored, plus any other ide local information like history of opened files

The issue is that Theia tries to store the state in `beforeunload` event which is not designed for it. Instead Theia should be refactored to move storing the state each time when the page becomes passive (loosing focus) or hidden (visibility change to hidden) and never in `beforeunload` event. After such refactoring Theia can be hidden (to store state) and then removed  from the page when the workspace goes to the stopping phase (to avoid `beforeunload` event).

### PR DESCRIPTION
**Pull request by akosyakov**
- [x] /werft ws-feature-flags=registry_facade
- [x] /werft https

#### What it does

- it is a follow-up of https://github.com/gitpod-io/gitpod/pull/1963 bumping up code to commit which integrates the supervisor
  - registration: https://github.com/gitpod-io/vscode/blob/b7d1abfe52328f9787543518c0d1698a798d9b28/src/vs/code/browser/workbench/workbench.html#L5-L7
  - implementation of IDEService:  https://github.com/gitpod-io/vscode/blob/b7d1abfe52328f9787543518c0d1698a798d9b28/src/vs/code/browser/workbench/workbench.ts#L252-L272
- fix #1959 by asking the IDE service to stop which triggers preserving state and preventing dispatching `beforeunload` event when the workspace is stopped to avoid confirmation dialogs

#### How to test

- login
- test Theia:
  - start a regular workspace http://ak-fix-code.staging.gitpod-dev.com/#https://github.com/gitpod-io/go-gin-app
    - change the layout, do dirty changes, try to reload the page, you should still see the confirmation dialog, cancel it and then stop the workspace
    - start the workspace from the loading screen, the layout should be preserved, but you should not see the confirmation dialog about dirty changes
  - start a prebuild workspace http://ak-fix-code.staging.gitpod-dev.com/#prebuild/https://github.com/gitpod-io/go-gin-app
    - ide code should be loaded, but the application should not be started, i.e. browser logs should be clean, no contributions or plugins loading
- test Code:
  - grant `devops` role to your user, i.e. run a query like: `update d_b_user set rolesOrPermissions = '["admin","devops"]' where name = 'akosyakov';`
  - start workspace http://ak-fix-code.staging.gitpod-dev.com/#https://github.com/akosyakov/go-gin-app
  - check that Code is loaded
  - click on Gitpod in remote status bar item (left most) and select stop workspace, check that loading screen is shown
  - restart the workspace from the loading screen and check that Code is loaded again

### CODE REVIEW

**Comment by akosyakov** [comment]
Now Code is too eager to access the supervisor:
Screenshot 2020-10-10 at 15 38 18

File: components/gitpod-protocol/src/ide-service.ts [hunk 1]
```diff
@@ -5,10 +5,20 @@
  */
 
 import { Event } from "./util/event";
+import { Disposable } from "./util/disposable";
 
-export type IDEState = 'init' | 'ready';
+export type IDEState = 'init' | 'ready' | 'terminated';
 
 export interface IDEService {
```
Comment (csweichel) [inline components/gitpod-protocol/src/ide-service.ts:12]: It's a bit late, I know. But how about we renamed this to `IDEFrontendService`?
Every time I see this interface I expect it to be some service on the backend side of things (this is the first "service" on the frontend - some Gitpod API mixings notwithstanding)

Comment (akosyakov) [reply components/gitpod-protocol/src/ide-service.ts:12]: actually had the same idea yesterday, thought that it is a rather the shape of `IDEFrontendApplication`

**Comment by csweichel** [comment]
The Theia part worked as expected. When I tried code, it did not start properly:
image

Looks to me like a timing issue. Code seems to be open before the content was ready.

**Comment by csweichel** [comment]
while stopping, I still saw bits of code:
image

**Comment by akosyakov** [comment]
@csweichel Could you review again?

It seems that there are still some http requests can slip before IDE ready, but it is not only for Code and there are not critical like icons. I will improve it later with support of service workers: https://github.com/gitpod-io/gitpod/issues/1982

**Review by csweichel — APPROVED** [review]
works as advertised

**Pull request merged**